instrumentation/spring/spring-integration-4.1/library/src/main/java/io/opentelemetry/instrumentation/spring/integration/TracingChannelInterceptor.java

```diff
@@ -0,0 +1,128 @@
+  private final Instrumenter<MessageWithChannel, Void> instrumenter;
+
+  TracingChannelInterceptor(
+      ContextPropagators propagators, Instrumenter<MessageWithChannel, Void> instrumenter) {
+    this.propagators = propagators;
+    this.instrumenter = instrumenter;
+  }
+
+  @Override
+  public Message<?> preSend(Message<?> message, MessageChannel messageChannel) {
+    Context parentContext = Context.current();
+    MessageWithChannel messageWithChannel = MessageWithChannel.create(message, messageChannel);
+
+    if (!instrumenter.shouldStart(parentContext, messageWithChannel)) {
+      return message;
+    }
+    Context context = instrumenter.start(parentContext, messageWithChannel);
+
+    MessageHeaderAccessor messageHeaderAccessor = createMutableHeaderAccessor(message);
+    propagators
+        .getTextMapPropagator()
+        .inject(context, messageHeaderAccessor, MessageHeadersSetter.INSTANCE);
+    messageHeaderAccessor.setHeader(CONTEXT_AND_SCOPE_KEY, ContextAndScope.makeCurrent(context));
+    return createMessageWithHeaders(message, messageHeaderAccessor);
+  }
+
+  @Override
+  public void postSend(Message<?> message, MessageChannel messageChannel, boolean sent) {}
+
+  @Override
+  public void afterSendCompletion(
+      Message<?> message, MessageChannel messageChannel, boolean sent, Exception e) {
+    ContextAndScope contextAndScope =
+        message.getHeaders().get(CONTEXT_AND_SCOPE_KEY, ContextAndScope.class);
+    if (contextAndScope != null) {
+      contextAndScope.close();
+      MessageWithChannel messageWithChannel = MessageWithChannel.create(message, messageChannel);
+      instrumenter.end(contextAndScope.getContext(), messageWithChannel, null, e);
+    }
+  }
+
+  @Override
+  public boolean preReceive(MessageChannel messageChannel) {
+    return true;
+  }
+
+  @Override
+  public Message<?> postReceive(Message<?> message, MessageChannel messageChannel) {
+    return message;
+  }
+
+  @Override
+  public void afterReceiveCompletion(
+      Message<?> message, MessageChannel messageChannel, Exception e) {}
+
+  @Override
+  public Message<?> beforeHandle(
+      Message<?> message, MessageChannel channel, MessageHandler handler) {
+    MessageWithChannel messageWithChannel = MessageWithChannel.create(message, channel);
+    Context context =
+        propagators
+            .getTextMapPropagator()
+            .extract(Context.current(), messageWithChannel, MessageHeadersGetter.INSTANCE);
+    Scope scope = context.makeCurrent();
+    MessageHeaderAccessor messageHeaderAccessor = MessageHeaderAccessor.getMutableAccessor(message);
+    messageHeaderAccessor.setHeader(SCOPE_KEY, scope);
+    return createMessageWithHeaders(message, messageHeaderAccessor);
+  }
+
+  @Override
+  public void afterMessageHandled(
+      Message<?> message, MessageChannel channel, MessageHandler handler, Exception ex) {
+    Scope scope = message.getHeaders().get(SCOPE_KEY, Scope.class);
```

Comment:
same question about `beforeHandle` and `afterMessageHandled`

Reply:
Spring Javadocs:

> Invoked inside the Runnable submitted to the Executor after calling the target MessageHandler regardless of the outcome (i.e. Exception raised or not) thus allowing for proper resource cleanup.

In addition to that, these two methods actually seem to be sort of deprecated in spring integration/cloud-stream, as they're only used once in one `MessageChannel` implementation from spring-messaging that is not used anywhere in the spring code.